### PR TITLE
Update the hometask #2

### PR DESCRIPTION
**Pull request by AndreyKovalevskiy**
Added generics so that the stack (Stack and StackStatistics classes) works with different types

### CODE REVIEW

File: Lesson2/SecondHomework.playground/Contents.swift [hunk 1]
```diff
@@ -32,15 +45,21 @@ class StackStatistics: Stack {
...
+    override func pop() -> T? {
         if !minimalValues.isEmpty {
             minimalValues.removeLast()
         }
         return super.pop()
         
     }
     
-    func minimumElement() -> Int? {
+    /// Returns the last minimum element in the stack. Returns nil when the invoking stack is empty.
```
Comment (BongDiDong) [inline Lesson2/SecondHomework.playground/Contents.swift]:
```suggestion
    /// Returns the minimum element in the stack. Returns nil when the invoking stack is empty.
```
what does that mean "the last minimum element"? It's confusing a bit.

**Review by BongDiDong — COMMENTED** [review]
Good work! Minor improvements:

File: Lesson2/SecondHomework.playground/Contents.swift [hunk 2]
```diff
@@ -32,15 +45,21 @@ class StackStatistics: Stack {
...
     
-    override func pop() -> Int? {
+    /// Removes and returns the top element of the stack. Returns nil when the invoking stack is empty.
+    /// - Complexity: O(1)
+    /// - Returns: The top element of the stack or nil.
+    override func pop() -> T? {
         if !minimalValues.isEmpty {
             minimalValues.removeLast()
         }
         return super.pop()
```
Comment (BongDiDong) [inline Lesson2/SecondHomework.playground/Contents.swift]:
```suggestion

```

Comment (AndreyKovalevskiy) [reply Lesson2/SecondHomework.playground/Contents.swift]: Thanks. I deleted the extra line

File: Lesson2/SecondHomework.playground/Contents.swift (hunk 1 above)
Comment (AndreyKovalevskiy) [reply Lesson2/SecondHomework.playground/Contents.swift]: Thanks. Corrected the description

**Review by Hannabolotova — APPROVED** [review]
Good job!

**Review by Iweinrazvadovskaya — APPROVED** [review]
Nice work ‎✓

**Review by BongDiDong — APPROVED** [review]
LGTM.

File: Lesson2/SecondHomework.playground/Contents.swift [hunk 3]
```diff
@@ -32,15 +45,20 @@ class StackStatistics: Stack {
         super.push(element: element)
     }
     
-    override func pop() -> Int? {
+    /// Removes and returns the top element of the stack. Returns nil when the invoking stack is empty.
+    /// - Complexity: O(1)
+    /// - Returns: The top element of the stack or nil.
+    override func pop() -> T? {
         if !minimalValues.isEmpty {
```
Comment (dersim-davaod) [inline Lesson2/SecondHomework.playground/Contents.swift]: Please read Array's documentation and try to simplify the implementation of `pop()`.

File: Lesson2/SecondHomework.playground/Contents.swift [hunk 4]
```diff
@@ -1,29 +1,42 @@
...
             return nil
         } else {
             return array.removeLast()
         }
     }
     
-    func top() -> Int? {
+    /// Returns the top element of the stack. Returns nil when the invoking stack is empty
+    /// - Complexity: O(1)
+    /// - Returns: The top element of the stack or nil.
```
Comment (dersim-davaod) [inline Lesson2/SecondHomework.playground/Contents.swift]: `Returns: The top element of the stack or nil`.

Is that means it returns `nil` randomly?

Comment (dersim-davaod) [reply Lesson2/SecondHomework.playground/Contents.swift]: Here and above.

File: Lesson2/SecondHomework.playground/Contents.swift [hunk 5]
```diff
@@ -1,29 +1,42 @@
...
-/// An object, which is a list of elements with a type integer, organized according to the LIFO principle.
-class StackStatistics: Stack {
-    private var minimalValues: [Int] = []
+/// An object, which is a list of elements, organized according to the LIFO principle.
+/// Each stack element must conform to the Comparable protocol.
+class StackStatistics<T: Comparable>: Stack<T> {
+    private var minimalValues: [T] = []
     
-    override func push(element: Int) {
+    /// Adds element on the top of the stack
```
Comment (dersim-davaod) [inline Lesson2/SecondHomework.playground/Contents.swift]: Please explain why you duplicate the same doc comments as for `super.push()`?

Comment (dersim-davaod) [reply Lesson2/SecondHomework.playground/Contents.swift]: Did you read about DRY principle? Please review the rest of doc comments.

File: Lesson2/SecondHomework.playground/Contents.swift [hunk 6]
```diff
@@ -32,15 +45,20 @@ class StackStatistics: Stack {
...
+    override func pop() -> T? {
         if !minimalValues.isEmpty {
             minimalValues.removeLast()
         }
         return super.pop()
-        
     }
     
-    func minimumElement() -> Int? {
+    /// Returns the minimum element in the stack. Returns nil when the invoking stack is empty.
```
Comment (dersim-davaod) [inline Lesson2/SecondHomework.playground/Contents.swift]:
```suggestion
    /// Returns the minimum element in the stack.
```

Typically you describe such details in `Returns` section.

File: Lesson2/SecondHomework.playground/Contents.swift [hunk 7]
```diff
@@ -1,29 +1,42 @@
-/// An object, which is a list of elements with a type integer, organized according to the LIFO principle.
-class Stack {
-    private var array: [Int] = []
+/// An object, which is a list of elements, organized according to the LIFO principle.
```
Comment (dersim-davaod) [inline Lesson2/SecondHomework.playground/Contents.swift]: I see: `/// An object, which is...`.

Is class an object?

**Review by dersim-davaod — CHANGES_REQUESTED** [review]
Consider improvements.

File: Lesson2/SecondHomework.playground/Contents.swift (hunk 3 above)
Comment (AndreyKovalevskiy) [reply Lesson2/SecondHomework.playground/Contents.swift]: made changes to the pop() method

File: Lesson2/SecondHomework.playground/Contents.swift (hunk 7 above)
Comment (AndreyKovalevskiy) [reply Lesson2/SecondHomework.playground/Contents.swift]: corrected class description

File: Lesson2/SecondHomework.playground/Contents.swift (hunk 6 above)
Comment (AndreyKovalevskiy) [reply Lesson2/SecondHomework.playground/Contents.swift]: corrected the description to the methods

File: Lesson2/SecondHomework.playground/Contents.swift (hunk 4 above)
Comment (AndreyKovalevskiy) [reply Lesson2/SecondHomework.playground/Contents.swift]: corrected the description to the methods

File: Lesson2/SecondHomework.playground/Contents.swift (hunk 5 above)
Comment (AndreyKovalevskiy) [reply Lesson2/SecondHomework.playground/Contents.swift]: Deleted the description of the push() and pop() methods in the StackStatistics class

**Review by dersim-davaod — COMMENTED** [review]
Everything is almost fine!

The last point is: some time ago I shared a link to https://swift.org/documentation/api-design-guidelines/. Please answer the question: when we have to document the complexity of properties?

**Comment by AndreyKovalevskiy** [comment]
@dersim-davaod , in API Design Guidelines writed: 

```Document the complexity of any computed property that is not O(1).```

Because my properties (array and minimalValues) are not computed I don’t documentation their complexity .

**Review by dersim-davaod — APPROVED** [review]
LGTM

**Pull request merged**